src/hotspot/share/gc/g1/g1ParScanThreadState.cpp

```diff
@@ -558,10 +558,7 @@ void G1ParScanThreadStateSet::flush() {
 
   for (uint worker_id = 0; worker_id < _n_workers; ++worker_id) {
     G1ParScanThreadState* pss = _states[worker_id];
-
-    if (pss == NULL) {
-      continue;
-    }
+    assert(pss != nullptr, "must be initialized");
```

Comment:
Is it guaranteed that all workers are initialized? As you state in the bug, they are lazily initialized, so I wonder if there are use cases where a worker would not get initialized during the collection. Something like, very little work and a lot of workers. 

If there is a guarantee here, does it hold below in `record_unused_optional_region()` as well? We are using the same construct with `continue`there.

Reply:
In `G1CollectedHeap::evacuate_initial_collection_set`:

```cpp
G1ParScanThreadStateSet per_thread_states(this,
                                          &rdcqs,
                                          workers()->active_workers(), <---- setting up states for workers
                                          collection_set()->young_region_length(),
                                          collection_set()->optional_region_length());
...
evacuate_initial_collection_set(&per_thread_states, may_do_optional_evacuation);
\
 -- task_time = run_task_timed(&g1_par_task); <--- launching all workers
```
Therefore, I think the assertion is safe.

> does it hold below in record_unused_optional_region() as well?

I think so.

Reply:
Changed; and tier1-3 pass.